text_generation/causal_lm/cpp/greedy_sampling.hpp

```diff
@@ -0,0 +1,165 @@
+    float penalty) {
+    if (penalty < 0) {
+        std::cout << "penalty must be a positive float, but got " << penalty;
+        return;
+    }
+    const float inv_penalty = 1.f / penalty;
+    const int vocab_size = last - first;
+    std::vector<bool> occurrence(vocab_size, false);
+    for (const int64_t id : input_ids) {
+        if (!occurrence[id]) {
+            first[id] *= (first[id] > 0) ? inv_penalty : penalty;
+        }
+        occurrence[id] = true;
+    }
+}
+
+void sampling_temperature(float* first, float* last, float temp) {
+    const float inv_temp = 1.f / temp;
+    for (float* it = first; it != last; it++) {
+        *it *= inv_temp;
+    }
+}
+
+struct SamplingParameters {
+    std::vector<int64_t> prompt;
+    int top_k = 0;
+    float top_p = 0.7;
+    float temp = 0.95;
+    float repeat_penalty = 1.1;
+    bool do_sample = true;
+};
+
+// GreedySampling processes logits prduced by a language model and chooses the token with
+// the highest probablity as the next token in the sequence. get_out_token() returns token 
+// ids selected by the algorithm. The value is used for next inference. 
+struct GreedySampling {
+    SamplingParameters parameters;
+    GreedySampling(SamplingParameters parameters) : parameters{ std::move(parameters) } {        
+    }
+
+    int64_t get_out_token(float* logits, size_t vocab_size) {
+        int64_t out_token;
+        std::vector<int64_t> prompt{ parameters.prompt };
+
+        // logits pre-process
+        if (parameters.repeat_penalty != 1.f) {
+            sampling_repetition_penalty(logits, logits + vocab_size, prompt, parameters.repeat_penalty);
+        }
+
+        if (parameters.do_sample)
+        {
+            if (parameters.temp > 0) {
+                sampling_temperature(logits, logits + vocab_size, parameters.temp);
+            }
+
+            std::vector<TokenIdScore> token_scores(vocab_size);
+            for (int i = 0; i < vocab_size; i++) {
+                token_scores[i] = TokenIdScore(i, logits[i]);
+            }
+
+            // top_k sampling
+            if (0 < parameters.top_k && parameters.top_k < (int)token_scores.size()) {
+                sampling_top_k(token_scores.data(), token_scores.data() + parameters.top_k,
+                    token_scores.data() + token_scores.size());
+                token_scores.resize(parameters.top_k);
+            }
+
+            // top_p sampling
+            if (0.f < parameters.top_p && parameters.top_p < 1.f) {
+                auto pos = sampling_top_p(token_scores.data(), token_scores.data() + token_scores.size(), parameters.top_p);
+                token_scores.resize(pos - token_scores.data());
+            }
+
+            // sample next token
+            sampling_softmax_inplace(token_scores.data(), token_scores.data() + token_scores.size());
+            for (size_t i = 0; i < token_scores.size(); i++) {
+                logits[i] = token_scores[i].score;

```

Comment:
You must take the values from logits here, not scores, and then do the softmax on them. The softmax already modified the scores in the top_p call, so the second softmax will skew the distribution towards the most probable token.

Reply:
I refer to chatglm cpp(https://github.com/li-plus/chatglm.cpp/blob/main/chatglm.cpp#L825) to implement this function. 
Is this implementation different from HuggingFace.

Reply:
I checked the softmax effects on the logits and the results for the llama model are that the second softmax makes the distribution nearly uniform and the third one makes the distribution exactly uniform. Here is how the probabilities of the most probable token are evolving:
```
Softmax 0 times: tensor([[ 0.0000,  8.6805,  9.4036,  8.4245, 10.8393]])
Softmax 1 times: tensor([[3.1250e-05, 2.7906e-01, 2.0102e-01, 7.2127e-02, 2.6164e-01]])
Softmax 2 times: tensor([[3.1250e-05, 4.1308e-05, 3.8207e-05, 3.3586e-05, 4.0595e-05]])
Softmax 3 times: tensor([[3.1250e-05, 3.1250e-05, 3.1250e-05, 3.1250e-05, 3.1250e-05]])
For uniform distribution: 3.125e-05
```
The second softmax makes model predictions useless. 

Huggingface doesn't transform the logits during `top_p`, just calculates softmax and filters the original logits based on the result. So `sampling_softmax_inplace` is not needed inside the `sampling_top_p` function.


Code for the softmax probs:
```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer


model_checkpoint = "JackFram/llama-68m"
hf_tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)
hf_model = AutoModelForCausalLM.from_pretrained(model_checkpoint)
texts = ["This is a test"]
tokenized = hf_tokenizer(texts, return_tensors="pt")

with torch.no_grad():
    t = hf_model(**tokenized).logits
    print("Softmax 0 times:", torch.max(t, dim=-1).values)
    for i in range(1, 4):
        t = torch.nn.functional.softmax(t, dim=-1,)
        print(f"Softmax {i} times:", torch.max(t, dim=-1).values)

print(f"For uniform distribution: {1 / hf_tokenizer.vocab_size}")
```

Reply:
@apaniukov Thanks a lot!

Reply:
I re-implemented the sampling_top_p function, the implementation is similar to the implementation of huggingface.